src/java.base/share/classes/jdk/internal/vm/VMSupport.java

```diff
@@ -167,4 +179,404 @@ public static int encodeThrowable(Throwable throwable, long buffer, int bufferSi
+                    writeLength(dos, array.length);
+                    for (short item : array) {
+                        dos.writeShort(item);
+                    }
+                } else if (componentType == boolean.class) {
+                    boolean[] array = (boolean[]) value;
+                    dos.writeByte('[');
+                    dos.writeByte('Z');
+                    writeLength(dos, array.length);
+                    for (boolean b : array) {
+                        dos.writeBoolean(b);
+                    }
+                } else if (componentType == String.class) {
+                    String[] array = (String[]) value;
+                    dos.writeByte('[');
+                    dos.writeByte('s');
+                    writeLength(dos, array.length);
+                    for (String s : array) {
+                        dos.writeUTF(s);
+                    }
+                } else if (componentType == Class.class) {
+                    Class<?>[] array = (Class<?>[]) value;
+                    dos.writeByte('[');
+                    dos.writeByte('c');
+                    writeLength(dos, array.length);
+                    for (Class<?> aClass : array) {
+                        dos.writeUTF(aClass.getName());
+                    }
+                } else if (componentType.isEnum()) {
+                    Enum<?>[] array = (Enum<?>[]) value;
+                    dos.writeByte('[');
+                    dos.writeByte('e');
+                    dos.writeUTF(componentType.getName());
+                    writeLength(dos, array.length);
+                    for (Enum<?> anEnum : array) {
+                        dos.writeUTF(anEnum.name());
+                    }
+                } else if (componentType.isAnnotation()) {
+                    Annotation[] array = (Annotation[]) value;
+                    dos.writeByte('[');
+                    dos.writeByte('@');
+                    writeLength(dos, array.length);
+                    for (Annotation annotation : array) {
+                        encodeAnnotation(dos, annotation);
+                    }
+                } else {
+                    dos.writeByte('x');
+                    dos.writeUTF(value.toString());
+                }
+
+            } else {
+                dos.writeByte('x');
+                dos.writeUTF(value.toString());
+            }
+        }
+    }
+
+    /**
+     * Helper for {@link #decodeAnnotations(byte[], AnnotationDecoder)} to convert a byte
+     * array (ostensibly produced by {@link VMSupport#encodeAnnotations}) into objects.
+     *
+     * @param <T> type to which a type name is {@linkplain #resolveType(String) resolved}
+     * @param <A> type of the object representing a decoded annotation
+     * @param <E> type of the object representing a decoded enum constant
+     * @param <X> type of the object representing a decoded error
+     */
+    public interface AnnotationDecoder<T, A, E, X> {
```

Comment:
I think it would be better to include some bound on the type parameters to better capture their intention
A extends java.lang.Annotatoin
E extends java.lang.Enum
etc.

Reply:
These types are *alternatives* to `java.lang.Annotation`, `java.lang.Enum` etc. That's the primary motivation for this PR, i.e. to be able to represent annotations without having to reify them as `java.lang.Annotation` objects.